src/hash_builder/mod.rs

```diff
@@ -227,33 +249,34 @@ impl HashBuilder {
 
             // Adjust the state masks for branch calculation
             let extra_digit = current[len];
-            if self.state_masks.len() <= len {
-                let new_len = len + 1;
-                trace!(target: "trie::hash_builder", new_len, old_len = self.state_masks.len(), "scaling state masks to fit");
-                self.state_masks.resize(new_len, TrieMask::default());
+
+            let new_len = len + 1;
+            if state_len < new_len {
+                trace!(target: "trie::hash_builder", new_len, old_len = state_len, "scaling state masks to fit");
+                self.state_masks.resize(new_len, *trie_default);
             }
             self.state_masks[len] |= TrieMask::from_nibble(extra_digit);
+
             trace!(
                 target: "trie::hash_builder",
                 ?extra_digit,
                 state_masks = ?self.state_masks,
             );
 
             // Adjust the tree masks for exporting to the DB
-            if self.tree_masks.len() < current.len() {
-                self.resize_masks(current.len());
+            if self.tree_masks.len() < current_len {
+                self.resize_masks(current_len);
             }
 
-            let mut len_from = len;
-            if !succeeding.is_empty() || preceding_exists {
-                len_from += 1;
-            }
+            let len_from = if s_and_p { new_len } else { len };
             trace!(target: "trie::hash_builder", "skipping {len_from} nibbles");
 
             // The key without the common prefix
             let short_node_key = current.slice(len_from..);
             trace!(target: "trie::hash_builder", ?short_node_key);
 
+            let current_slice = &current.slice(..len_from);
```

Comment:
We will never use this variable twice, because the places where it's used are `if !build_extensions { ... }` and `if build_extension && ... { ... }`